store/plugin-commands-server/src/start.ts

```diff
@@ -118,6 +118,11 @@ export async function start (
   // Set fd to null so we only attempt to close it once.
   fd = null
   await close(fdForClose)
+
+  // Intentionally avoid returning control back to the caller until the server
+  // exits. This defers cleanup operations that should not run before the server
+  // finishes.
+  await server.waitForClose
```

Comment:
The main pnpm command handler will wait for this promise to resolve here:

https://github.com/pnpm/pnpm/blob/131c723030ec4016274ca1cef33aadd66eb0e3b9/pnpm/src/main.ts#L283

Everything after this `main.ts` line looks safe to run on server close. I could use another reviewer to gut check this change in behavior though.